### ISSUE TITLE
GWR().predict() method: How are parameters estimated?

### ISSUE THREAD

**Issue by thordur-orn** [pysal/mgwr#124]
Does someone have a formula or an explanation for how we estimate the parameters (betas) for the test set? Like when we use the predict() method of the GWR, which parameters are being used?

**Comment by veager** [comment]
I have the similar question. How does the out-of-sample prediction work, specially for those locations not existing in the train dataset?

According to the principle of GWR, every location is estimated by an local LR model. A new location doesn't have an estimated LR model, but how to the prediction?

**Comment by ljwolf** [comment]
The algorithm is quite simple, we just have not had the time to implement it.

1. Pool the input training data and prediction data together. 
2. Using the kernel function and estimated bandwidth, construct the smoothed design matrix at out of sample points from the pooled data. Also, construct the smoothed Y for each prediction site
3. Compute the local betas using the normal equations for the model form
4. Multiply the out of sample X and estimated out of sample local betas to get your prediction. 

We would really love to have this implemented, so if you manage it, definitely contribute back!

**Comment by veager** [comment]
Thanks @ljwolf reply! I have looked through these codes and roughly understood this process. 

But I still confuse about the process of "smoothed Y". Does that mean use some smoothes methods to estimate Y for the new coming out-of-bag samples first and then fit the GWR as usually? But I didn't find any function to smooth that Y?

Is there any paper/textbook/tutorial to explain this process of out-of-sample prediction in detail?

Thanks so much!

**Comment by ljwolf** [comment]
> Does that mean use some smoothes methods to estimate Y for the new coming out-of-bag samples first and then fit the GWR as usually?

Yes. This is only necessary to construct the local betas for the out of bag estimates. You don't re-fit the entire GWR, only estimate the local modes on the out of bag points using the smoothed X from the pooled data and smoothed y from the training data alone. 

> examples

I'm not aware of any, but maybe @TaylorOshan or @Ziqi-Li might?

**Comment by ljwolf** [comment]
And to be sure, this algorithm will only work with GWR. I don't think this is suitable for MGWR, because of the process specific bandwidths

**Comment by TaylorOshan** [comment]
I believe there actually is a predict function that is currently
implemented, but there is a bug that only allows it to be used when the
training set is larger than the testing set. As you noted, even this will
only work for GWR though and not for MGWR.
